src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py

```diff
@@ -220,6 +225,145 @@ def __init__(
+                    is_any_parent_unique_node = any(
+                        node_count_map[n] == 1 for n in node_parents_map[node]
+                    )
+                    if is_any_parent_unique_node:
+                        return True
+                    else:
+                        has_multi_parents = len(node_parents_map[node]) > 1
+                        if has_multi_parents:
+                            return True
+                return False
+
+            traverse(root)
+            return {node for node in node_count_map if is_duplicate_subtree(node)}
+
+        # if there is no duplicate node, no optimization will be performed
+        duplicate_plan_set = find_duplicate_subtrees(self)
+        if not duplicate_plan_set:
+            return self
+
+        plan_to_query_map = {}
+        duplicate_plan_to_cte_map = {}
+        duplicate_plan_to_table_name_map = {}
+
+        def build_plan_to_query_map_in_post_order(node: SnowflakePlan) -> None:
+            """
+            Builds a mapping from query plans to queries that are optimized with CTEs,
+            in post-traversal order. We can get the final query from the mapping value of the root node.
+            The reason of using poster-traversal order is that chained CTEs have to be built
+            from bottom (innermost subquery) to top (outermost query).
+            """
+            if not node.source_plan or node in plan_to_query_map:
+                return
+
+            for child in node.source_plan.children:
+                build_plan_to_query_map_in_post_order(child)
+
+            if not node.placeholder_query:
+                plan_to_query_map[node] = node.queries[-1].sql
+            else:
+                plan_to_query_map[node] = node.placeholder_query
+                for child in node.source_plan.children:
+                    # replace the placeholder (id) with child query
+                    plan_to_query_map[node] = plan_to_query_map[node].replace(
+                        child._id, plan_to_query_map[child]
+                    )
+
+            # duplicate subtrees will be converted CTEs
+            if node in duplicate_plan_set:
+                # when a subquery is converted a CTE to with clause,
+                # it will be replaced by `SELECT * from TEMP_TABLE` in the original query
+                table_name = random_name_for_temp_object(TempObjectType.TABLE)
+                select_stmt = project_statement([], table_name)
+                duplicate_plan_to_table_name_map[node] = table_name
+                duplicate_plan_to_cte_map[node] = plan_to_query_map[node]
+                plan_to_query_map[node] = select_stmt
+
+        build_plan_to_query_map_in_post_order(self)
+
+        # construct with clause
+        with_stmt = cte_statement(
+            list(duplicate_plan_to_cte_map.values()),
+            list(duplicate_plan_to_table_name_map.values()),
+        )
+        final_query = with_stmt + SPACE + plan_to_query_map[self]
```

Comment:
since we are finding the topmost duplicate node, we don't need to concern ourselves with finding the correct order, right. This is cool

Reply:
Yeah, because we build `duplicate_plan_to_cte_map` in post order, where children node is ahead of parent node. It satisfies the requirement of with clause:

> A WITH clause can refer recursively to itself, and to other CTEs that appear earlier in the same clause. For instance, cte_name2 can refer to cte_name1 and itself, while cte_name1 can refer to itself, but not to cte_name2.